plugins/producer_plugin/test/test_snapshot_scheduler.cpp

```diff
@@ -0,0 +1,141 @@
+         return e.to_detail_string().find("Block spacing exceeds defined by start and end range") != std::string::npos;
+      });
+
+      producer_plugin::snapshot_request_information sri_start_end = {.block_spacing = 1000, .start_block_num = 50000, .end_block_num = 5000};
+      BOOST_CHECK_EXCEPTION(scheduler.schedule_snapshot(sri_start_end), invalid_snapshot_request, [](const fc::assert_exception& e) {
+         return e.to_detail_string().find("End block number should be greater or equal to start block number") != std::string::npos;
+      });
+   }
+   {
+      boost::filesystem::path temp = boost::filesystem::temp_directory_path() / boost::filesystem::unique_path();
+
+      try {
+         std::promise<std::tuple<producer_plugin*, chain_plugin*>> plugin_promise;
+         std::future<std::tuple<producer_plugin*, chain_plugin*>> plugin_fut = plugin_promise.get_future();
+
+         std::thread app_thread([&]() {
+            fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::debug);
+            std::vector<const char*> argv =
+                  {"test", "--data-dir", temp.c_str(), "--config-dir", temp.c_str(),
+                   "-p", "eosio", "-e", "--max-transaction-time", "475", "--disable-subjective-billing=true"};
+            appbase::app().initialize<chain_plugin, producer_plugin>(argv.size(), (char**) &argv[0]);
+            appbase::app().startup();
+            plugin_promise.set_value(
+                  {appbase::app().find_plugin<producer_plugin>(), appbase::app().find_plugin<chain_plugin>()});
+            appbase::app().exec();
+         });
+
+         auto [prod_plug, chain_plug] = plugin_fut.get();
+         std::deque<block_state_ptr> all_blocks;
+         std::promise<void> empty_blocks_promise;
+         std::future<void> empty_blocks_fut = empty_blocks_promise.get_future();
+         auto ab = chain_plug->chain().accepted_block.connect([&](const block_state_ptr& bsp) {
+            static int num_empty = std::numeric_limits<int>::max();
+            all_blocks.push_back(bsp);
+            if(bsp->block->transactions.empty()) {
+               --num_empty;
+               if(num_empty == 0) empty_blocks_promise.set_value();
+            } else {// we want a few empty blocks after we have some non-empty blocks
+               num_empty = 10;
+            }
+         });
+         auto pp = appbase::app().find_plugin<producer_plugin>();
+         auto bs = chain_plug->chain().block_start.connect([&pp](uint32_t bn) {
+            // catching pending snapshot
+            auto& pending = pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots;
+            if (pending && pending->size()==1) {
+               // lets check the head block num of it, it should be 8 + 1 = 9
+               // this means we are getting a snapshot for correct block # as well
+               BOOST_CHECK_EQUAL(9, pending->begin()->head_block_num);   
+            }
```

Comment:
This could happen if block_start is called before the snapshots are scheduled, of if we schedules only one non-recurring snapshot.

```suggestion
            if (!pp->get_snapshot_requests().snapshot_requests.empty()) {
               auto& pending = pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots;
               if (pending && pending->size()==1) {
                  // lets check the head block num of it, it should be 8 + 1 = 9
                  // this means we are getting a snapshot for correct block # as well
                  BOOST_CHECK_EQUAL(9, pending->begin()->head_block_num);   
               }
            }
```

Reply:
agreed, done